toolchain/check/import_cpp.cpp

```diff
@@ -18,20 +18,57 @@
 #include "toolchain/diagnostics/format_providers.h"
 
 namespace Carbon::Check {
+namespace {
+
+auto GenerateCppIncludesHeaderFilename(llvm::StringRef importing_file_path)
+    -> std::string {
+  return llvm::Twine(importing_file_path)
+      .concat(".generated.cpp_imports.h")
+      .str();
+}
+
+auto GenerateCppIncludesHeaderCode(
+    llvm::ArrayRef<std::pair<llvm::StringRef, SemIRLoc>> imports)
+    -> std::string {
+  constexpr llvm::StringLiteral IncludePrefix(R"(#include ")");
+  constexpr llvm::StringLiteral IncludeSuffix(R"("
+)");
+
+  std::string code;
+  for (const auto& [path, _] : imports) {
+    code.append(
+        llvm::Twine(IncludePrefix).concat(path).concat(IncludeSuffix).str());
+  }
+  return code;
```

Comment:
I'd find these easier to read as escaped characters instead of raw string literals; `")"` hides the closing `"` for me, for example, and the newline breaks the indentation of code. That said, had you looked at inlining the constant values?

I think we often do things like this with streams, such as:

```suggestion
  std::string code;
  llvm::raw_string_ostream code_str(code);
  for (const auto& [path, _] : imports) {
    code_str << "#include \"" << path << "\"\n";
  }
  return code;
```

(ed: dropping `Twine` suggestion that doesn't work)

Reply:
While I dropped a twine suggestion, I should still note the twine use as-is is allocating a string per-line; I think the ostream approach might also be more efficient as a consequence. It'd probably be slightly more efficient with buffer reserving, but at a certain point this isn't performance critical and I think it's more readable like this.

Does make me want to do an owning string_ostream wrapper, though. Maybe I'll take a moment and finally do a non-test version of that.

Reply:
Thanks, this makes perfect sense!
I tried different approaches but for some reason didn't consider streams.